Allow certificate expiration checks no matter what ssl/tls verification settings are

scott-shields-github opened the pull request:
### What does this PR do?
This change allows for the retrieval of a certificate when `tls_verify` is `False`. Since we have the certificate, it can be checked for expiration whether or not we are connecting with ssl/tls verification. Issue #13526 

Related support ticket: https://help.datadoghq.com/hc/en-us/requests/1036304

### Motivation
Since users are able to disable ssl/tls verification for various reasons, there should also be a way to check for certificate expiration in those circumstances. For organizations using self-signed certificates, this allows for checking expiration without needing to copy those certs to all Datadog agent instances.

### Additional Notes
The core of this change only changes the format of the retrieved certificate. [SSLSocket.getpeercert](https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.getpeercert), when used in the default manner (`binary_form=False`), returns a dict representation of the certificate if the certificate is validated. If not validated, like when ssl/tls validation is disabled, an empty dict is returned (causing the exceptions seen with `tls_verify: False`). Using `binary_form=True`, the DER encoded form of the certificate is returned whether or not the certificate is validated. This allows it to be checked for an expiration regardless of ssl/tls verification without changing anything else outside of the retrieval format of the certificate.

There are a few other small changes for testing so configs/instances would be unique for each test.

### Review checklist (to be filled by reviewers)

- [x] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [x] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [x] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [x] PR must have `changelog/` and `integration/` labels attached
- [x] If the PR doesn't need to be tested during QA, please add a `qa/skip-qa` label.

yzhan289 reviewed (commented):
Overall I like this change, although I think this would be considered a breaking change, so I'll have some more eyes on it before approving

yzhan289 commented:
Hi @scott-shields-github, after talking with the team, we think it would be best to add a config option to achieve this so this wouldn't be a breaking change for existing customers. Could you update the PR so it keeps the original implementation but allows you to retrieve the cert regardless even with `tls_verify: False`?

scott-shields-github commented on http_check/tests/compose/certs-builder/Dockerfile:
```diff
@@ -21,7 +21,7 @@ RUN openssl genrsa -out valid.mock.key 2048 \
 ## Generate a self-signed non-expired certificate for selfsigned.mock
 
-RUN openssl req -nodes -new -x509 -days 1 -keyout selfsigned.mock.key -out selfsigned.mock.cert -subj "/C=US/ST=NY/O=foo/CN=selfsigned.mock"
+RUN openssl req -nodes -new -x509 -days 20 -keyout selfsigned.mock.key -out selfsigned.mock.cert -subj "/C=US/ST=NY/O=foo/CN=selfsigned.mock"
```
This was just arbitrarily longer than the 10 days used in the legacy tests. This just allowed me to keep those consistent.

scott-shields-github replied:
I've updated this to 200 to match the other valid cert

scott-shields-github commented:
Hey @yzhan289...I can do that. 

Just to confirm, the previous behavior with `tls_verify: False`, an `Empty or no certificate found.` exception and an `AgentCheck.UNKNOWN` response, is the desired behavior for that setting and we'll have a new, separate configuration property to override that behavior and retrieve the certificate?

yzhan289 commented:
> Just to confirm, the previous behavior with `tls_verify: False`, an `Empty or no certificate found.` exception and an `AgentCheck.UNKNOWN` response, is the desired behavior for that setting and we'll have a new, separate configuration property to override that behavior and retrieve the certificate?

Correct. Although it's raising an exception currently, `http_check` is a fairly popular integration and we want to make sure we maintain backwards compatibility for this change. You can add a section in the [Troubleshooting](https://github.com/DataDog/integrations-core/tree/master/http_check#troubleshooting) section of the `http_check` README to make your change more explicit in case others want to have the same behavior.

scott-shields-github commented on http_check/tests/test_http_integration.py:
```diff
@@ -183,11 +230,10 @@ def test_check_tsl_ca_cert(aggregator, dd_run_check):
     with mock.patch('ssl.SSLSocket.getpeercert', side_effect=Exception()):
         # Run the check for the one instance configured with days left
-        http_check = HTTPCheck('', {}, [CONFIG_EXPIRED_SSL['instances'][0]])
-        dd_run_check(http_check)
```
I've changed this back and mocked the cert path.

scott-shields-github commented:
I've added the new option and added a few more tests to make sure the initial behavior is maintained. I've also added the new option to the docs. I don't seem to be able to add a label for the changelog label check failure.

buraizu commented on http_check/README.md:
```diff
@@ -46,6 +46,7 @@ See the [sample http_check.d/conf.yaml][3] for a full list and description of av
 | `include_content`                | When set to `true`, the check includes the first 500 characters of the HTTP response body in notifications. The default value is `false`.                                                                                                        |
 | `collect_response_time`          | By default, the check collects the response time (in seconds) as the metric `network.http.response_time`. To disable, set this value to `false`.                                                                                                 |
 | `tls_verify`                     | Instructs the check to validate the TLS certificate of services when reaching to `url`.                                                                                                                                                          |
+| `tls_retrieve_non_validated_cert`| If `tls_verify` is set to `false`, setting this to `true` will allow the certificate to be examined for an expiration date.                                                                                                                      |
```
```suggestion
| `tls_retrieve_non_validated_cert`| If `tls_verify` is set to `false`, setting this to `true` allows the certificate to be examined for an expiration date.                                                                                                                      |
```

buraizu replied:
Minor suggestion per our [contributing guidelines](https://github.com/DataDog/documentation/blob/master/CONTRIBUTING.md#tense)

scott-shields-github replied:
updated...thanks for the link!

buraizu commented on http_check/datadog_checks/http_check/data/conf.yaml.example:
```diff
@@ -136,6 +136,13 @@ instances:
     # check_certificate_expiration: true
 
+    ## @param tls_retrieve_non_validated_cert - boolean - optional - default: false
+    ## When set to true, this (optional) option allows certificates to be retrieved from a peer
+    ## whether or not "tls_verify" is set to true or false. This allows the certificate to be
```
```suggestion
    ## When set to true, this option allows certificates to be retrieved from a peer
    ## whether or not `tls_verify` is set to true or false. This allows the certificate to be
```

buraizu reviewed (changes requested):
Thanks for the PR, left some minor feedback on the doc update

buraizu reviewed (dismissed):
Thanks for implementing those changes, LGTM!

yzhan289 commented on http_check/datadog_checks/http_check/http_check.py:
```diff
@@ -302,18 +329,27 @@ def check_cert_expiration(self, instance, timeout, instance_ca_certs):
+            # To maintain backwards compatability, if we aren't validating tls/certs, do not process
+            # the returned binary cert unless specifically configured to with tls_retrieve_non_validated_cert
+            if (
+                instance.get("tls_verify") is False and not instance.get("tls_retrieve_non_validated_cert")
```
```suggestion
                instance.get("tls_verify") is False and is_affirmative(instance.get("tls_retrieve_non_validated_cert", False))
```

scott-shields-github replied:
I committed this by an accident from this comment and am working on fixing it up now. What is the best way to import this function? I'm getting errors with the linux environment with import failures while attempting to import from `datadog_checks_base.datadog_checks.base`.

scott-shields-github replied:
Please disregard this...this was my error :)

scott-shields-github replied:
This has been updated and tests are passing now

yzhan289 commented on http_check/README.md:
```diff
@@ -46,6 +46,7 @@ See the [sample http_check.d/conf.yaml][3] for a full list and description of av
+| `tls_retrieve_non_validated_cert`| If `tls_verify` is set to `false`, setting this to `true` allows the certificate to be examined for an expiration date.                                                                                                                      |
 | `tls_ignore_warning`             | If `tls_verify` is set to `true`, it disables any security warnings from the SSL connection.                                                                                                                                                     |
 | `tls_ca_cert`                    | This setting allows you to override the default certificate path as specified in `init_config`                                                                                                                                                   |
 | `check_certificate_expiration`   | When `check_certificate_expiration` is enabled, the service check checks the expiration date of the SSL certificate. **Note**: This causes the SSL certificate to be validated, regardless of the value of the `tls_verify` setting. |
```
Small nit, could we move your `tls_retrieve_non_validated_cert` under this `check_certificate_expiration` option since the two are related? I would also update your description to include that `check_certificate_expiration` should be set to true to make the connection explicit.

scott-shields-github replied:
Will do

scott-shields-github replied:
updated

yzhan289 commented on http_check/assets/configuration/spec.yaml:
```diff
@@ -113,6 +113,14 @@ files:
         example: true
+    - name: tls_retrieve_non_validated_cert
+      description: |
+        When set to true, this option allows certificates to be retrieved from a peer
```
```suggestion
        When set to true along with enabling `check_certificate_expiration`, this option allows certificates to be retrieved from a peer
```

scott-shields-github replied:
updated

yzhan289 commented on http_check/datadog_checks/http_check/http_check.py:
```diff
@@ -302,18 +329,28 @@ def check_cert_expiration(self, instance, timeout, instance_ca_certs):
+            # To maintain backwards compatability, if we aren't validating tls/certs, do not process
+            # the returned binary cert unless specifically configured to with tls_retrieve_non_validated_cert
+            if (
+                instance.get("tls_verify") is False
```
```suggestion
                not is_affirmative(instance.get("tls_verify", True))
```
Oops I forgot to leave this comment earlier

scott-shields-github replied:
updated
